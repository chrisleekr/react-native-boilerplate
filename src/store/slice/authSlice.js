import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';

import authService from '../../service/authService';
import TokenService from '../../service/tokenService';
import {getErrorMessages} from '../../util/requestUtils';

export const login = createAsyncThunk(
  'auth/login',
  async (args, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().auth;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    const {username, password} = args;

    try {
      const response = await authService.login({username, password});

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const register = createAsyncThunk(
  'auth/register',
  async (args, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().auth;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    const {email, username, password, firstName, lastName} = args;

    try {
      const response = await authService.register({
        email,
        username,
        password,
        firstName,
        lastName
      });

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const passwordResetRequest = createAsyncThunk(
  'auth/passwordResetRequest',
  async (args, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().auth;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    const {email} = args;

    try {
      const response = await authService.passwordResetRequest({
        email
      });

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

const initialState = {
  loading: 'idle',
  currentRequestId: undefined,
  errors: [],
  isAuthenticated: false,
  accessToken: '',
  refreshToken: ''
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    setState(state, action) {
      const {key, value} = action.payload;
      state[key] = value;
    },
    initialise(state) {
      TokenService.loadToken(jwt => {
        if (jwt) {
          state.accessToken = jwt.accessToken;
          state.refreshToken = jwt.refreshToken;
          state.isAuthenticated = true;
        }
      });
    },
    logout(state) {
      state.errors = [];
      state.isAuthenticated = false;
      state.accessToken = '';
      state.refreshToken = '';

      TokenService.resetToken();
    },
    resetErrors(state) {
      state.errors = [];
    }
  },
  extraReducers: builder => {
    builder
      .addCase(login.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
          state.isAuthenticated = false;
        }
      })
      .addCase(login.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];

          const {auth_key: accessToken, refresh_auth_key: refreshToken} =
            action.payload.data;

          state.accessToken = accessToken;
          state.refreshToken = refreshToken;
          state.isAuthenticated = true;

          TokenService.updateToken({
            accessToken: accessToken,
            refreshToken: refreshToken
          });
        }
      })
      .addCase(login.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
        state.isAuthenticated = false;
      })
      .addCase(register.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(register.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(register.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      })
      .addCase(passwordResetRequest.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(passwordResetRequest.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(passwordResetRequest.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      });
  }
});

export const {setState, initialise, logout, resetErrors} = authSlice.actions;

const {reducer: authReducer} = authSlice;
export default authReducer;
