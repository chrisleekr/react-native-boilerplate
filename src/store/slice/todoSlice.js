import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';

import todoService from '../../service/todoService';
import {getErrorMessages} from '../../util/requestUtils';

export const list = createAsyncThunk(
  'todo/list',
  async (_, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().todo;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    try {
      const response = await todoService.list();

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const addOne = createAsyncThunk(
  'todo/addOne',
  async ({name, note, state}, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().todo;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    try {
      const response = await todoService.postOne({name, note, state});

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const updateOne = createAsyncThunk(
  'todo/updateOne',
  async ({todo}, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().todo;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    try {
      const response = await todoService.patchOne(todo);

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const deleteOne = createAsyncThunk(
  'todo/deleteOne',
  async ({todo}, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().todo;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    try {
      const response = await todoService.deleteOne(todo);

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

const initialState = {
  loading: 'idle',
  currentRequestId: undefined,
  errors: [],
  todoOngoing: [],
  todoCompleted: []
};

const todoSlice = createSlice({
  name: 'todo',
  initialState,
  reducers: {
    setState(state, action) {
      const {key, value} = action.payload;
      state[key] = value;
    },
    resetErrors(state) {
      state.errors = [];
    }
  },
  extraReducers: builder => {
    builder
      .addCase(list.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(list.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];

          const {rows: todo} = action.payload.data;

          state.todoOngoing = todo.filter(t => t.state === 'ongoing');
          state.todoCompleted = todo.filter(t => t.state === 'completed');
        }
      })
      .addCase(list.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      })
      .addCase(addOne.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(addOne.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(addOne.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      })
      .addCase(updateOne.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(updateOne.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(updateOne.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      })
      .addCase(deleteOne.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(deleteOne.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(deleteOne.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      });
  }
});

export const {setState, resetErrors} = todoSlice.actions;

const {reducer: todoReducer} = todoSlice;
export default todoReducer;
