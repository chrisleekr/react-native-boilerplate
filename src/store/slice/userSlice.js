import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';

import userService from '../../service/userService';
import {getErrorMessages} from '../../util/requestUtils';

export const me = createAsyncThunk(
  'user/me',
  async (_, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().user;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    try {
      const response = await userService.me();
      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

export const updateMe = createAsyncThunk(
  'user/updateMe',
  async (args, {getState, requestId, rejectWithValue}) => {
    const {currentRequestId, loading} = getState().user;

    if (loading !== 'pending' || requestId !== currentRequestId) {
      return;
    }

    const {email, password, firstName, lastName} = args;

    try {
      const response = await userService.updateMe({
        email,
        password,
        firstName,
        lastName
      });

      return response;
    } catch (e) {
      if (!e.response) {
        throw e;
      }

      return rejectWithValue(e.response);
    }
  }
);

const initialState = {
  loading: 'idle',
  currentRequestId: undefined,
  errors: [],
  user: {}
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    setState(state, action) {
      const {key, value} = action.payload;
      state[key] = value;
    },
    resetErrors(state) {
      state.errors = [];
    }
  },
  extraReducers: builder => {
    builder
      .addCase(me.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(me.fulfilled, (state, action) => {
        const {requestId} = action.meta;

        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];

          const user = action.payload.data;

          state.user = user;
        }
      })
      .addCase(me.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      })
      .addCase(updateMe.pending, (state, action) => {
        if (state.loading === 'idle') {
          state.loading = 'pending';
          state.currentRequestId = action.meta.requestId;
          state.errors = [];
        }
      })
      .addCase(updateMe.fulfilled, (state, action) => {
        const {requestId} = action.meta;
        if (
          state.loading === 'pending' &&
          state.currentRequestId === requestId
        ) {
          state.loading = 'idle';
          state.currentRequestId = undefined;
          state.errors = [];
        }
      })
      .addCase(updateMe.rejected, (state, action) => {
        state.loading = 'idle';
        state.errors = getErrorMessages(action.payload);
        state.currentRequestId = undefined;
      });
  }
});

export const {setState, resetErrors} = userSlice.actions;

const {reducer: userReducer} = userSlice;
export default userReducer;
