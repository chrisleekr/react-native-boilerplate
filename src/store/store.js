import {configureStore} from '@reduxjs/toolkit';
import {combineReducers} from 'redux';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER
} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Axios
import setupInterceptors from '../service/setupInterceptors';

// Slices
import authReducer from './slice/authSlice';
import todoReducer from './slice/todoSlice';
import userReducer from './slice/userSlice';

const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage
};

const rootReducer = combineReducers({
  auth: authReducer,
  todo: todoReducer,
  user: userReducer
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  // middleware option needs to be provided for avoiding the error. ref: https://redux-toolkit.js.org/usage/usage-guide#use-with-redux-persist
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      // Since axios is throwing serializecheck error, so had to turn off.
      // serializableCheck: false
      // But let see how it goes with following list instead of turnning it off.
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'auth/login/rejected',
          'auth/register/rejected',
          'auth/passwordResetRequest/rejected',
          'todo/list/rejected',
          'todo/addOne/rejected',
          'todo/updateOne/rejected',
          'todo/deleteOne/rejected',
          'user/me/rejected',
          'user/updateMe/rejected'
        ]
      }
    })
});

setupInterceptors(store);

export const persistor = persistStore(store);
export default store;
