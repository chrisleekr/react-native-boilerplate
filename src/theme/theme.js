import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
