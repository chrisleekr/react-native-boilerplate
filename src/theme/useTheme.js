/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
