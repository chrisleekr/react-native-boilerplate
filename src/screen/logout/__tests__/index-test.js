import React from 'react';
import * as redux from 'react-redux';
import {cleanup, render, waitFor} from '@testing-library/react-native';

import {ThemeProvider} from '../../../theme/useTheme';

import store from '../../../store/store';

import LogoutScreen from '../index';

let rendered;

const mockLogout = {type: 'LOGOUT'};

import * as authSlice from '../../../store/slice/authSlice';
const spyLogout = jest.spyOn(authSlice, 'logout');

const component = (
  <ThemeProvider>
    <redux.Provider store={store}>
      <LogoutScreen />
    </redux.Provider>
  </ThemeProvider>
);

describe('LogoutScreen', () => {
  afterEach(cleanup);

  beforeEach(() => {
    spyLogout.mockImplementation(() => mockLogout);
  });

  beforeEach(async () => {
    rendered = render(component);

    await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
  });

  it('matches snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('triggers logout', () => {
    expect(spyLogout).toHaveBeenCalled();
  });
});
