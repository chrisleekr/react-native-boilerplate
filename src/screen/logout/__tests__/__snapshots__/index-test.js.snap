// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutScreen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#f0f0f0",
      },
      undefined,
    ]
  }
  testID="logout-screen"
>
  <Text>
    You have been logged out successfully.
  </Text>
</RNCSafeAreaView>
`;
